### ISSUE TITLE
running quarkus multimodule project  in dev mode leads to Ambiguous dependencies error

### ISSUE THREAD

**Issue by NorthFuture** [quarkusio/quarkus#3687]
**Describe the bug**
I have a multi module maven project
testmultimodule-root
  | testmultimodule-runner (depends on testmultimodule-presentation-adapter, quarkus-resteasy)
  | testmultimodule-presentation-adapter ( depends on testmultimodule-service-adapter,quarkus-resteasy+quarkus-maven-plugin)
  | testmultimodule-service-adapter

if I package the application (mvn package) and run the resulting jar (java -jar testmultimodule-runner-1.0-SNAPSHOT-runner.jar) everything works fine, 

if I run the package in dev mode (mvn clean compile quarkus:dev:) i get an ambigous dependency error 

Caused by: javax.enterprise.inject.spi.DeploymentException: javax.enterprise.inject.AmbiguousResolutionException: Ambiguous dependencies for type org.testmultimodule.GreetingServ
ice and qualifiers [@Default]                                                                                                                                                     
        - java member: org.testmultimodule.GreetingResource#service                                                                                                               
        - declared on CLASS bean [types=[org.testmultimodule.GreetingResource, java.lang.Object], qualifiers=[@Default, @Any], target=org.testmultimodule.GreetingResource]       
        - available beans:                                                                                                                                                        
                - CLASS bean [types=[org.testmultimodule.GreetingService, java.lang.Object], qualifiers=[@Default, @Any], target=org.testmultimodule.GreetingService]             
                - CLASS bean [types=[org.testmultimodule.GreetingService, java.lang.Object], qualifiers=[@Default, @Any], target=org.testmultimodule.GreetingService]             

If I run  -DnoDeps the application starts up, but hotreload doesn't work for testmultimodule-presentation-adapter or testmultimodule-service-adapter

It seems testmultimodule-service-adapter is loaded twice in two different classloaders.

**To Reproduce**
1. unzip the archive
[test-multimodule.zip](https://github.com/quarkusio/quarkus/files/3538281/test-multimodule.zip)

2. from testmultimodule-root, mvn clean compile quarkus:dev:

**Environment (please complete the following information):**
 - Microsoft Windows [Versione 10.0.18362.295]
 - openjdk version "1.8.0_222"
   OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_222-b10)
   OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.222-b10, mixed mode)
 - N/A
 - Quarkus version: 0.21.1

**Comment by geoand** [comment]
@mkouba @manovotn perhaps you would like to take a look?

**Comment by manovotn** [comment]
This isn't really coming from Arc - when Arc gets its hands on list of all `ApplicationArchive`s (used [here](https://github.com/quarkusio/quarkus/blob/master/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/BeanArchiveProcessor.java#L34)), there is already a duplicate.

Looking at how this is created,[ I can see that at certain point, Quarkus is looking for marker files (`beans.xml` in this case) using `ClassLoader.getResource("META-INF/beans.xml)`](https://github.com/quarkusio/quarkus/blob/master/core/deployment/src/main/java/io/quarkus/deployment/index/ApplicationArchiveBuildStep.java#L177). And here what we get is, among other results:
* `file:/home/somePath/testmultimodule-service-adapter/target/classes/META-INF/beans.xml`
* `file:/home/somePath/test/testmultimodule-root/../testmultimodule-service-adapter/target/classes/META-INF/beans.xml`

E.g. the very same file twice, only once we reach it via parent module.
Note that all of the above is what I observed *without* using `-DnoDeps`.
This part of Quarkus is outside of my expertise though, maybe @aloubyansky would know more?

**Comment by aloubyansky** [comment]
We should add this to our testsuite.

**Comment by InfoSec812** [comment]
I am having a similar issue with dev mode on a multimodule project. I have the following modules structure:

todo:
  - jsclient
  - javaclient
  - server
  - implementation
  - frontend

The `jsclient`, `server`, and `javaclient` are all generated using the OpenAPI Generator, but I am using custom templates to ensure that they are correctly configured for Quarkus.

I then compile together `server` and `implementation` using Dev mode and I get errors like:

```
Caused by: java.lang.ClassCastException: com.redhat.quarkus.demo.todo.TodosApiServiceImpl_ClientProxy cannot be cast to com.redhat.demos.quarkus.openapi.todo.api.TodosApiService
        at com.redha
```

When I compile `implementation` to a runnable jar, it works fine. BTW, example project can be found HERE: https://github.com/InfoSec812/quarkus-openapi-todo-demo/tree/switch-to-maven-multimodule

**Comment by aloubyansky** [comment]
@NorthFuture I just tried your reproducer with the recently released 1.3.0.CR2 and it worked.
@InfoSec812 I tried yours as well and got a different error then what you reported (see below). Perhaps I did something wrong. Could you please try 1.3.0.CR2 yourself and if it still doesn't work open another issue?
Thanks a lot both of you for providing reproducers.
```
java.lang.RuntimeException: Could not find /home/aloubyansky/playground/quarkus-openapi-todo-demo/server/../openapi.yml on the classpath
    at io.swagger.v3.parser.util.ClasspathHelper.loadFileFromClasspath (ClasspathHelper.java:31)
    at io.swagger.v3.parser.OpenAPIV3Parser.readWithInfo (OpenAPIV3Parser.java:149)
    at io.swagger.v3.parser.OpenAPIV3Parser.readLocation (OpenAPIV3Parser.java:58)
    at io.swagger.parser.OpenAPIParser.readLocation (OpenAPIParser.java:16)
    at org.openapitools.codegen.config.CodegenConfigurator.toContext (CodegenConfigurator.java:451)
    at org.openapitools.codegen.config.CodegenConfigurator.toClientOptInput (CodegenConfigurator.java:497)
    at org.openapitools.codegen.plugin.CodeGenMojo.execute (CodeGenMojo.java:705)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
```

**Comment by InfoSec812** [comment]
@aloubyansky I'll try to get a look at it tonight.

**Comment by y-luis** [comment]
This issue seems inactive, but I am facing the same problem, is there any tip on how to avoid it?

**Comment by aloubyansky** [comment]
@y-luis please open a new issue with instructions how to reproduce the problem, a reproducer would be even better. Thanks.
